File: ddss/load.ts
import * as readline from "node:readline";
import { parse } from "atsds-bnf";
import { initializeDatabase, insertOrIgnore, Fact, Idea } from "./orm.js";
import { strRuleGetStrIdea } from "./utility.js";
import type { Sequelize } from "sequelize";

export async function main(addr: string, sequelize?: Sequelize) {
    if (!sequelize) {
        sequelize = await initializeDatabase(addr);
    }

    const rl = readline.createInterface({
        input: process.stdin,
        terminal: false,
    });

    try {
        for await (const line of rl) {
            const data = line.trim();
            if (data === "" || data.startsWith("//")) {
                continue;
            }

            try {
                const ds = parse(data);
                const dsStr = ds.toString();
                await insertOrIgnore(Fact, dsStr);
                const idea = strRuleGetStrIdea(dsStr);
                if (idea) {
                    await insertOrIgnore(Idea, idea);
                }
            } catch (e) {
                console.error(`error: ${e}`);
            }
        }
    } finally {
        rl.close();
    }
}

File: ddss/load.py
import asyncio
import sys
from sqlalchemy.ext.asyncio import async_sessionmaker, AsyncSession
from apyds_bnf import parse
from .orm import insert_or_ignore, Facts, Ideas
from .utility import str_rule_get_str_idea


async def main(session: async_sessionmaker[AsyncSession]) -> None:
    try:
        async with session() as sess:
            for line in sys.stdin:
                data = line.strip()
                if data == "" or data.startswith("//"):
                    continue

                try:
                    ds = parse(data)
                except Exception as e:
                    print(f"error: {e}")
                    continue

                await insert_or_ignore(sess, Facts, ds)
                if idea := str_rule_get_str_idea(ds):
                    await insert_or_ignore(sess, Ideas, idea)
            await sess.commit()
    except asyncio.CancelledError:
        pass
